### PR TITLE
JENKINS-31625

### PR DESCRIPTION
**Pull request by klimas7**

### CODE REVIEW

File: src/test/java/org/biouno/unochoice/TestParametersOrder.java [hunk 1]
```diff
@@ -60,12 +60,18 @@ public void testParametersOrder() {
...
         parameters.put("A", "A");
 
+        System.out.println("D1");
+        /*
         ChoiceParameter parameter = new ChoiceParameter(
                 "script001", "description", "random name", new GroovyScript(SCRIPT, FALLBACK_SCRIPT),
-                ChoiceParameter.PARAMETER_TYPE_MULTI_SELECT, true);
+                ChoiceParameter.PARAMETER_TYPE_MULTI_SELECT, true, 0);
+
+        System.out.println("D2");
```
Comment (kinow) [inline src/test/java/org/biouno/unochoice/TestParametersOrder.java]: We don't need to print anything in the console. And I think there was a problem with indentation below here too

**Comment by klimas7** [comment]
I'm sorry for that test :) I forgot revert a small 'debug  investigation'

**Comment by kinow** [comment]
You rock @klimas7! I'm testing this pull request locally now :-)

File: src/main/resources/org/biouno/unochoice/CascadeChoiceParameter/config.jelly [hunk 2]
```diff
@@ -55,4 +55,7 @@
   <f:entry title="${%Enable filters}" field="filterable" help="${rootURL}/../plugin/uno-choice/help-filterableParameters.html">
     <f:checkbox name="parameter.filterable" checked="${instance.filterable}">${%Filterable}</f:checkbox>
   </f:entry>
+  <f:entry title="${%Start filtered length}" field="filterLength" help="${rootURL}/../plugin/uno-choice/help-filterLengthParameters.html">
```
Comment (kinow) [inline src/main/resources/org/biouno/unochoice/CascadeChoiceParameter/config.jelly]: We may need to think a better description, though I don't have one. @imoutsatsos ? Maybe something like "Letters needed to active filtering". But we can do that after this gets merged, so nothing to change here @klimas7

File: src/main/resources/org/biouno/unochoice/stapler/unochoice/unochoice.js [hunk 3]
```diff
@@ -636,6 +644,10 @@ var UnoChoice = UnoChoice || (function($) {
             var filterElement = _self.getFilterElement();
             var filteredElement = _self.getParameterElement();
             var text = filterElement.value.toLowerCase();
+            if (text.length != 0 && text.length < _self.getFilterLength()) {
+                console.log("Filter pattern to short " + text.length + " < " + _self.getFilterLength());
```
Comment (kinow) [inline src/main/resources/org/biouno/unochoice/stapler/unochoice/unochoice.js]: While testing, noticed that I got one message like this for each character entered. Could you remove this line @klimas7 ? In the future, we can discuss a way to enable/disable via Jenkins UI debug in the JS console. But for now let's limit what we put there (we are going to reduce even more I believe for JENKINS-34188).

File: src/main/webapp/help-filterLengthParameters.html [hunk 4]
```diff
@@ -0,0 +1,5 @@
+<div>
+    <p>
+        The length of the pattern of the filter when the filter starts to work.
```
Comment (kinow) [inline src/main/webapp/help-filterLengthParameters.html]: I also think we need to improve this help documentation. Maybe a simple example, but I can take care of that after merging :-)

File: src/main/resources/org/biouno/unochoice/CascadeChoiceParameter/config.jelly (hunk 2 above)
Comment (klimas7) [inline src/main/resources/org/biouno/unochoice/CascadeChoiceParameter/config.jelly]: "Letters needed to active filtering" this is better description. I changed it

**Comment by kinow** [comment]
Thanks for keeping the pull request up to date @klimas7 

Release 1.5.3 is almost ready. Posted here 

https://groups.google.com/forum/#!topic/biouno-developers/nJmUhz6HWQo

telling about the release, and about including this pull request in the next experimental release on Friday.

**Comment by klimas7** [comment]
Hi @kinow can you look at  build plugin on jenkins, builds are failing with outofmemory error:
[#108](https://jenkins.ci.cloudbees.com/job/plugins/job/active-choices-plugin/108/console), [#109](https://jenkins.ci.cloudbees.com/job/plugins/job/active-choices-plugin/109/console), [#110](https://jenkins.ci.cloudbees.com/job/plugins/job/active-choices-plugin/110/console), [#112](https://jenkins.ci.cloudbees.com/job/plugins/job/active-choices-plugin/112/console)

**Comment by kinow** [comment]
Hmmm, looks like powermockito is now playing very nice with the rest of the test harness. Will take a look this week to see if I can change anything in the tests, otherwise will have to ask Jenkins infra guys.

But at least I can confirm for now that the build is passing on my environment, no failing tests.

**Comment by kinow** [comment]
Got the same PermGen errors in another computer, with Java 7. After some research, confirmed it was PowerMock. Found one issue with part of the solution: https://github.com/jayway/powermock/issues/294

Applied locally, tests passing with no PermGen or warning. Pushed. Let's wait and see now :-)

**Comment by kinow** [comment]
By the way, I reformatted the tests, and thus the conflicting files. I can merge them manually, so no need to update the pull request. My bad.

**Comment by kinow** [comment]
FWIW, added a note in the Jenkins Wiki for this PermGen errors: https://wiki.jenkins-ci.org/display/JENKINS/Mocking+in+Unit+Tests

And here's the life saver issue with the annotation for fixing another error related to the crypto package that I also had in my local environment: https://github.com/jayway/powermock/issues/294

**Comment by klimas7** [comment]
Hi, @kinow 
I fixed the "perm gen error". Problem wasn't a powermock but parallel launch tests.
If you started the five tests parallel you started five jenkins instances in the same time 
I set the "forkCount" at 0 in maven-Surefire plugin

**Comment by kinow** [comment]
Pull request merged. Will be released with 1.5.4. Watch the JIRA in order to receive an update when the version has been released.

Thanks, and sorry for the delay on merging it. Cherry-picked the commits, massaged the code a bit, and tested with a simple project. Looking good :-)

**Pull request closed without merging**